### ISSUE TITLE
ci flake: kafka-exactly-once-sink.td:317:1: error: missing record 0: Record {

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#16778]
### What version of Materialize are you using?

9a0e0e0cad

### How did you install Materialize?

Docker image

### What is the issue?

https://buildkite.com/materialize/tests/builds/47179#018533ea-61a4-4bd8-ba82-271c384fb8b5

succeeded on retry. I have not seen that one myself before personally.

The container logs do not seem to show any Kafka or Mz error.

@bki

### Relevant log output

```source
Verifying results in Kafka topic rt-binding-progress-test-output-931503860
Received error from Kafka stream consumer: deadline has elapsed
kafka-exactly-once-sink.td:317:1: error: missing record 0: Record {
    headers: [],
    key: None,
    value: Some(
        {
            (
                "before",
                Union {
                    index: 0,
                    inner: Null,
                    n_variants: 2,
                    null_variant: Some(
                        0,
                    ),
                },
            ),
            (
                "after",
                Union {
                    index: 1,
                    inner: {
                        (
                            "a",
                            Long(1),
                        ),
                        (
                            "b",
                            Long(1),
                        ),
                    },
                    n_variants: 2,
                    null_variant: Some(
                        0,
                    ),
                },
            ),
        },
    ),
}
     |
 316 |
 317 | $ kafka-verify-data format=avro sink=materialize.public.rt_binding_progress_test_sink
     | ^
```

**Comment by aljoscha** [comment]
@bkirwi could you please take a look?

**Comment by bkirwi** [comment]
Odd! Here's the most interesting bit of the logs (u147 is the relevant sink):

```
testdrive-materialized-1     | environmentd: 2022-12-21T09:06:01.311931Z  INFO mz_service::grpc: GrpcClient /tmp/environmentd-mzcompose-test-00000000-0000-0000-0000-000000000000-0/storage-u147/controller-0: connected
[...]
testdrive-schema-registry-1  | [2022-12-21 09:08:01,404] INFO 172.16.48.6 - - [21/Dec/2022:09:08:01 +0000] "GET /subjects/rt-binding-progress-test-output-931503860-key/versions/latest HTTP/1.1" 404 99 "-" "-" GETsT (io.confluent.rest-utils.requests)
[...]
testdrive-materialized-1     | storage-u147: 2022-12-21T09:08:01.437644Z  INFO mz_storage::sink::kafka: kafka-u147: initial as_of: Antichain { elements: [1671613561017] }, latest progress record: None
[progress messges on other sinks]
testdrive-materialized-1     | storage-u147: 2022-12-21T09:08:01.463412Z  INFO mz_storage::sink::kafka: kafka-u147: sending progress for gate ts: 1671613563999
[progress messages on other sinks]
testdrive-materialized-1     | environmentd: 2022-12-21T09:08:01.703004Z  INFO mz_adapter::catalog: drop sink materialize.public.compaction_test_sink (u145)
[several more drops elided]
testdrive-materialized-1     | environmentd: 2022-12-21T09:08:01.703577Z  INFO mz_adapter::catalog: drop sink materialize.public.rt_binding_progress_test_sink (u147)
```

For some reason, many of the sinks were dropped right as the sink in question was getting started (but a couple of minutes after it was created). It's possible that this is testdrive's own cleanup, in which case it sounds like _something_ prevented the sink from starting up promptly. No smoking guns in the logs in that respect, though.

In any case this looks more like an "actual" timeout than the sink going wrong or dropping things, which is the less concerning class of bug. I don't think I'll have time to dig any deeper before EOY, but IMO this is only urgent to look into further if we start seeing more of these!

**Comment by philip-stoev** [comment]
The issue now readily happens in Nightly , in the kafka-matrix test. 

If you run it like this:

```
cd test/kafka-matrix
while bash -c -e "./mzcompose down -v ; ./mzcompose run default" ;  do : ; done
```

it will happen within an hour or two.

Once the issue happens, no amount of waiting will fix it -- the topic in question is simply empty, as seen when reading it with a consumer:

```
docker exec -it kafka-matrix-schema-registry-1 kafka-avro-console-consumer --bootstrap-server=kafka:9092 --topic rt-binding-progress-test-output-XXXXXXXXXXXXXXXX --from-beginning
```

**Comment by aljoscha** [comment]
@bkirwi Could you please take another look?

**Comment by bkirwi** [comment]
The logs show the sink process starting and being connected to by environmentd at 2023-01-15T13:55:44.153383Z and dropped at  2023-01-15T13:57:44.726189Z, with nothing much in between. (It does get as far as initializing the producer, early on.)

It's plausible the fact that it never recovers is because the test framework drops the sink; we saw the same thing the first time this error popped up. Nevertheless, it would surprising for a sink to take almost two minutes to get started.

I've got this running on my scratch machine now with a little extra logging... we'll see if we can narrow down where it's getting hung.

**Comment by philip-stoev** [comment]
Keeps happening with regularity https://buildkite.com/materialize/nightlies/builds/1833#01866ecf-6a0a-49a8-93d1-622650b50f56

**Comment by def-** [comment]
@bkirwi This also happened *twice* with `persist_stats_audit_percent=100`: https://buildkite.com/materialize/nightlies/builds/2396#01882a85-d1d1-41b4-88b0-38be65c237a4
Maybe that makes it easier to reproduce?

**Comment by philip-stoev** [comment]
It happens very reliably with pubsub enabled, but I guess it is a separate issue since the record does get emitted 10 minutes later - https://github.com/MaterializeInc/materialize/issues/19328

**Comment by bkirwi** [comment]
Yeah, likely unrelated -- this error is effectively the Kafka version of "query timed out", so it's unfortunately likely that all sorts of bugs would have this exact symptom. (And I note that Dennis' branch also has the pubsub stuff enabled... I'll see if I can repro without it.)

**Comment by bkirwi** [comment]
I was not able to reproduce this purely with stats-auditing enabled but pubsub enabled.

Perhaps this new version has to do with pubsub somehow, or another change that went in recently?

**Comment by def-** [comment]
It happens quite often in CI on main with different replica settings, see for example https://buildkite.com/materialize/nightlies/builds/2498#01886f01-54e9-49d8-bf55-df368f24bcb2

**Comment by bkirwi** [comment]
I (seem to have) fixed one source of Kafka timeouts in #19879. I don't have a good guess yet at whether this should fix most of the issues in CI, though it does at least seem to help a lot with the repro in the linked issue when running on a scratch machine. We'll see how the next couple weeks of CI look I guess!

**Comment by bkirwi** [comment]
Nobody has complained of this in a long time, so I'll assume the linked PR sorted us out for now. Feel free to reopen if that's wrong!
